src/orderbook/state.ts

```diff
@@ -101,23 +172,82 @@ export class AccountState {
             validUntil: parseInt(data.validUntil),
             priceNumerator: BigInt(data.priceNumerator),
             priceDenominator: BigInt(data.priceDenominator),
+            remainingAmount: BigInt(data.priceDenominator),
           })
           break
+        case 'SolutionSubmission':
+          /* eslint-disable no-case-declarations */
+          const submitter = data.submitter
+          const burntFees = BigInt(data.burntFees)
+          this.updateBalance(submitter, 0, amount => amount + burntFees)
+          lastSolution = { submitter, burntFees }
+          break
         case 'TokenListing':
           this.addToken(parseInt(data.id), data.token)
           break
-        case 'Deposit':
-        case 'SolutionSubmission':
         case 'Trade':
+          this.updateBalance(
+            data.owner,
+            parseInt(data.sellToken),
+            amount => amount - BigInt(data.executedSellAmount),
+          )
+          this.updateOrderRemainingAmount(
+            data.owner,
+            parseInt(data.orderId),
+            amount => amount - BigInt(data.executedSellAmount),
+          )
+          this.updateBalance(
+            data.owner,
+            parseInt(data.buyToken),
+            amount => amount + BigInt(data.executedBuyAmount),
+          )
+          break
         case 'TradeReversion':
+          if (lastSolution !== undefined) {
+            const { submitter, burntFees } = lastSolution
+            this.updateBalance(submitter, 0, amount => amount - burntFees)
+            lastSolution = undefined
+          }
+          this.updateBalance(
+            data.owner,
+            parseInt(data.sellToken),
+            amount => amount + BigInt(data.executedSellAmount),
+          )
+          this.updateOrderRemainingAmount(
+            data.owner,
+            parseInt(data.orderId),
+            amount => amount + BigInt(data.executedSellAmount),
+          )
+          this.updateBalance(
+            data.owner,
+            parseInt(data.buyToken),
+            amount => amount - BigInt(data.executedBuyAmount),
+          )
+          break
         case 'WithdrawRequest':
+          this.setPendingWithdrawal(
+            data.user,
+            data.token,
+            parseInt(data.batchId),
+            BigInt(data.amount),
+          )
+          break
         case 'Withdraw':
-          // TODO(nlordell): Handle more events.
+          this.updateBalance(
```

Comment:
ubernit: should we assert that the balance is at least sufficient and as high as the requested one?

Reply:
Currently we assert at the end that the balances are all positive, so if the withdraw would overdraw the user balance we would fail our assertion.

I plan on refactoring the giant `switch ... case` statement to be more "dumb" as you suggested and then include the balance on withdraw like you suggest.